runtime/near-evm-runner/src/utils.rs

```diff
@@ -183,23 +187,428 @@ pub fn format_log(topics: Vec<H256>, data: &[u8]) -> std::result::Result<Vec<u8>
+            return Err(VMLogicError::EvmError(EvmError::InvalidMetaTransactionMethodName));
+        }
+        let mut args = vec![];
+        let first = remains.chars().next().unwrap();
+        if is_arg_start(first) {
+            let (arg, r) = Arg::parse(remains)?;
+            remains = r;
+            args.push(arg);
+            while remains.chars().next() == Some(',') {
+                remains = consume(remains, ',')?;
+                let (arg, r) = Arg::parse(remains)?;
+                remains = r;
+                args.push(arg);
+            }
+        }
+
+        let remains = consume(remains, ')')?;
+
+        Ok((args, remains))
+    }
+}
+
+impl Method {
+    fn parse(method_name: &str) -> Result<(Method, &str)> {
+        let (name, remains) = parse_ident(method_name)?;
+        let (args, remains) = Arg::parse_args(remains)?;
+        Ok((
+            Method {
+                name,
+                args,
+                raw: method_name[..method_name.len() - remains.len()].to_string(),
+            },
+            remains,
+        ))
+    }
+}
+
+impl MethodAndTypes {
+    pub fn parse(method_name: &str) -> Result<Self> {
+        let mut method_name = method_name;
+        let mut parsed_types = HashMap::new();
+        let mut type_sequences = vec![];
+        let (method, mut types) = Method::parse(method_name)?;
+        while types.len() > 0 {
+            let (ty, remains) = Method::parse(types)?;
+            type_sequences.push(ty.name.clone());
+            parsed_types.insert(ty.name.clone(), ty);
+            types = remains;
+        }
+        Ok(MethodAndTypes { method, types: parsed_types, type_sequences })
+    }
+}
+
+fn parse_ident(text: &str) -> Result<(String, &str)> {
+    let mut chars = text.chars();
+    if text.len() == 0 || !is_arg_start(chars.next().unwrap()) {
+        return Err(VMLogicError::EvmError(EvmError::InvalidMetaTransactionMethodName));
+    }
+
+    let mut i = 1;
+    for c in chars {
+        if !is_arg_char(c) {
+            break;
+        }
+        i += 1;
+    }
+    Ok((text[..i].to_string(), &text[i..]))
+}
+
+fn parse_type_raw(text: &str) -> Result<(String, &str)> {
+    let i =
+        text.find(' ').ok_or(VMLogicError::EvmError(EvmError::InvalidMetaTransactionMethodName))?;
+    Ok((text[..i].to_string(), &text[i..]))
+}
+
+fn consume(text: &str, c: char) -> Result<&str> {
+    let first = text.chars().next();
+    if first.is_none() || first.unwrap() != c {
+        return Err(VMLogicError::EvmError(EvmError::InvalidMetaTransactionMethodName));
+    }
+
+    Ok(&text[1..])
+}
+
+fn is_arg_start(c: char) -> bool {
+    c.is_ascii_alphabetic() || c == '_'
+}
+
+fn is_arg_char(c: char) -> bool {
+    c.is_ascii_alphanumeric() || c == '_'
+}
+
+/// Return a signature of the method_name
+/// E.g. method_signature(Methods before parse: "adopt(uint256 petId)") -> "adopt(uint256)"
+fn methods_signature(method_and_type: &MethodAndTypes) -> String {
```

Comment:
It's not clear to me why we need this function. The [ABI](https://docs.soliditylang.org/en/develop/abi-spec.html) suffices to specify how EVM function selectors look.

Why do we accept non-standard function signatures? Where are they coming from?

I followed the call chain far enough to assume they will be externally provided as an argument to the `meta_call` method. Is that correct? Who will be calling this method?

Reply:
We need `adopt(uint256 petId)` format in doing eip-712 hash. It needs to first keccak("adopt(uint256 petId)") then pack petId as uint256 after it.

Reply:
@artob here is correct, i got from near-pet-shop was a old version where method name is given as "adopt(uint256 petId)" in message, that's why i needed to keccak("adopt(uint256 petId)"), right now our message is "adopt(uint256)" and thus these functions are unnecessary

Reply:
Deleted

Reply:
@artob I just found this cannot be replaced. this:
```
    let types = "NearTx(string evmId,uint256 nonce,uint256 feeAmount,address feeAddress,address contractAddress,string contractMethod,Arguments arguments)".to_string() + &arguments;
    bytes.extend_from_slice(&keccak(types.as_bytes()).as_bytes());

```
must have argument like **evmId** and also argument names in arguments" otherwise generated eip-712 hash doesn't match what metamask generated. For example I tried passing 
```
NearTx(string,uint256,uint256,address,address,string,Arguments)Arguments(uint256) and generated eip-712 hash doesn't match metamask's. You can run this to get metamask generated eip-712 hash:
https://gist.github.com/ailisp/0f65bc1a3827b03141ae6601918a67a2

I consult eip-712 spec, seems metamask is correct here, see:
![image](https://user-images.githubusercontent.com/13259400/102568225-e61f9180-4097-11eb-8aab-eefa5d3e8d01.png)

So, we should still pass method with method name as message (adopt(uint256 petId)) to correctly reproduce the signature, instead of adopt(uint256)

Reply:
@artob in these two commits:
https://github.com/near/nearcore/pull/3704/commits/90a6518603d61bcb7559a3153648d189647baf43
https://github.com/near/nearcore/pull/3704/commits/2c2b80d7ec6c7ca919e6519a27c29b6c7d7c3804
i link to spec and explain every use of method_sig vs method_def. Hope this can resolve the confusion.

Reply:
@ailisp There is still a significant confusion here. I think you are being misled by EIP-712's structure signature format (`NearTx(...)`) resembling the Solidity function selector signature format (`adopt(uint256)`) . However, there is actually no connection between the two, even if they do superficially resemble each other.

EIP-712 is about structure types, the ABI is about function calls. The structure signature includes the structure's property names; the function signature does _not_ include the function's parameter names, only their types. Does that make sense?

Reply:
> We need adopt(uint256 petId) format in doing eip-712 hash. It needs to first keccak("adopt(uint256 petId)") then pack petId as uint256 after it.

`adopt` is a function (or method, if you like) name, not a structure name. So, EIP-712 is not relevant to it. As per the Solidity ABI spec, to call a function `adopt(uint256 petId)` in the EVM, you take `keccak("adopt(uint256)")` (_without_ the parameter name!) and then append the argument to that. EIP-712 is about structures, not functions/methods.

Reply:
Yes, but, we need to pass this format:
> include the function's parameter names, append any type signatures for its non-atomic argument types.

in message, because this is the only place we have function types and type definitions that is later used in parse_meta_call to reproduce signature.

Reply:
OK, thanks for that. I will read the code again with this in mind :+1: